feat(wallet) extract json blobs and add custom migration support

stefandunca opened the pull request:
### Update status-desktop [#10746](https://github.com/status-im/status-desktop/pull/new/migrate_transfers_blobs-10746)

Extended the migration process with a generic way of applying custom migration code on top of the SQL files. This is needed to extract the status from the JSON blob receipt and store it in table

Extract and migrate existing status from receipt to the activity Extract the rest of the useful receipt data

Notes:

- there is already an index on transfers table, SQLite creates one for each unique constraint, therefore, add only status to a new index
- the planned refactoring and improvements to the database have been postponed due to time constraints. Got the time to migrate the data though; extracting it can be done later for a more efficient implementation

stefandunca committed:
feat(wallet) extract json blobs and add custom migration support

Extended the migration process with a generic way of applying custom
migration code on top of the SQL files. The implementation provides
a safer way to run GO code along with the SQL migrations and possibility
of rolling back the changes in case of failure to keep the database
consistent.
This custom GO migration is needed to extract the status from
the JSON blob receipt and store it in transfers table.

Other changes:
- Add NULL DB value tracking to JSONBlob helper
- Index status column on transfers table
- Remove unnecessary panic calls
- Move log_parser to wallet's common package and use to extract token
  identity from the logs

Notes:
- there is already an index on transfers table, sqlite creates one for
  each unique constraint therefore add only status to a new index
- the planned refactoring and improvements to the database have been
  postponed due to time constraints. Got the time to migrate the data
  though, extracting it can be done later for a more efficient
  implementation

Update status-desktop #10746

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (76)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | fa7ee8c4 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/1/) | 2023-05-28 10:52:09 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230528-104921-fa7ee8-pr3540.zip) |
| :heavy_multiplication_x: | fa7ee8c4 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/1/) | 2023-05-28 10:52:42 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/1//consoleText) |
| :heavy_check_mark: | fa7ee8c4 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/1/) | 2023-05-28 10:53:38 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230528-104921-fa7ee8-pr3540.zip) |
| :heavy_check_mark: | fa7ee8c4 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/1/) | 2023-05-28 10:54:44 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230528-104921-fa7ee8-pr3540.aar) |
| | | | | | | |
| :heavy_check_mark: | ea128cfc | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/2/) | 2023-05-28 21:46:08 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230528-214338-ea128c-pr3540.zip) |
| :heavy_check_mark: | ea128cfc | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/2/) | 2023-05-28 21:46:17 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230528-214341-ea128c-pr3540.zip) |
| :heavy_multiplication_x: | ea128cfc | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/2/) | 2023-05-28 21:47:02 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/2//consoleText) |
| :heavy_check_mark: | ea128cfc | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/2/) | 2023-05-28 21:47:18 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230528-214338-ea128c-pr3540.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 368c83bf | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/3/) | 2023-05-28 21:52:31 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/3//consoleText) |
| :heavy_check_mark: | 368c83bf | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/3/) | 2023-05-28 21:53:27 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230528-215123-368c83-pr3540.zip) |
| :heavy_check_mark: | 368c83bf | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/3/) | 2023-05-28 21:53:55 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230528-215123-368c83-pr3540.zip) |
| :heavy_check_mark: | 368c83bf | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/3/) | 2023-05-28 21:54:54 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230528-215123-368c83-pr3540.aar) |
| | | | | | | |
| :heavy_check_mark: | c3682afb | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/4/) | 2023-05-29 20:55:00 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230529-205220-c3682a-pr3540.zip) |
| :heavy_check_mark: | c3682afb | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/4/) | 2023-05-29 20:55:31 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230529-205220-c3682a-pr3540.zip) |
| :heavy_check_mark: | c3682afb | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/4/) | 2023-05-29 20:57:08 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230529-205215-c3682a-pr3540.aar) |
| :heavy_multiplication_x: | c3682afb | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/4/) | 2023-05-29 20:58:32 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/4//consoleText) |
| :heavy_multiplication_x: | c3682afb | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/5/) | 2023-05-29 21:40:12 | ~23 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 2710306b | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/5/) | 2023-05-30 09:06:34 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230530-090402-271030-pr3540.zip) |
| :heavy_check_mark: | 2710306b | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/5/) | 2023-05-30 09:06:47 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230530-090401-271030-pr3540.zip) |
| :heavy_check_mark: | 2710306b | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/5/) | 2023-05-30 09:07:46 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230530-090400-271030-pr3540.aar) |
| :heavy_check_mark: | 2710306b | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/6/) | 2023-05-30 09:14:39 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 306e6b59 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/6/) | 2023-05-30 12:12:34 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230530-120923-306e6b-pr3540.zip) |
| :heavy_check_mark: | 306e6b59 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/6/) | 2023-05-30 12:25:48 | ~16 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230530-120923-306e6b-pr3540.zip) |
| :heavy_check_mark: | 306e6b59 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/6/) | 2023-05-30 12:26:25 | ~17 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230530-120923-306e6b-pr3540.aar) |
| | | | | | | |
| :heavy_check_mark: | 30ab52ce | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/7/) | 2023-05-30 12:36:36 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230530-123339-30ab52-pr3540.zip) |
| :heavy_check_mark: | 30ab52ce | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/7/) | 2023-05-30 12:46:38 | ~12 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230530-123339-30ab52-pr3540.aar) |
| :heavy_check_mark: | 30ab52ce | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/7/) | 2023-05-30 12:47:11 | ~13 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230530-123339-30ab52-pr3540.zip) |
| :heavy_check_mark: | 30ab52ce | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/8/) | 2023-05-30 13:08:18 | ~23 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/8//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | c1eb3ec1 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/9/) | 2023-05-31 06:32:25 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/9//consoleText) |
| :heavy_check_mark: | c1eb3ec1 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/8/) | 2023-05-31 06:33:27 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230531-063038-c1eb3e-pr3540.zip) |
| :heavy_check_mark: | c1eb3ec1 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/8/) | 2023-05-31 06:34:10 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230531-063034-c1eb3e-pr3540.zip) |
| :heavy_check_mark: | c1eb3ec1 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/8/) | 2023-05-31 06:35:05 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230531-063034-c1eb3e-pr3540.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 1e3a63d0 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/10/) | 2023-05-31 19:54:53 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/10//consoleText) |
| :heavy_check_mark: | 1e3a63d0 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/9/) | 2023-05-31 19:54:54 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230531-195229-1e3a63-pr3540.zip) |
| :heavy_check_mark: | 1e3a63d0 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/9/) | 2023-05-31 19:55:53 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230531-195229-1e3a63-pr3540.zip) |
| :heavy_check_mark: | 1e3a63d0 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/9/) | 2023-05-31 19:57:10 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230531-195229-1e3a63-pr3540.aar) |
| | | | | | | |
| :heavy_check_mark: | d8b24ca1 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/10/) | 2023-06-02 11:05:21 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230602-110250-d8b24c-pr3540.zip) |
| :heavy_check_mark: | d8b24ca1 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/10/) | 2023-06-02 11:05:33 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230602-110250-d8b24c-pr3540.zip) |
| :heavy_multiplication_x: | d8b24ca1 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/11/) | 2023-06-02 11:05:43 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/11//consoleText) |
| :heavy_check_mark: | d8b24ca1 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/10/) | 2023-06-02 11:06:56 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230602-110250-d8b24c-pr3540.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 3dc96062 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/12/) | 2023-06-02 11:51:28 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/12//consoleText) |
| :heavy_check_mark: | 3dc96062 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/11/) | 2023-06-02 11:52:28 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230602-115016-3dc960-pr3540.zip) |
| :heavy_check_mark: | 3dc96062 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/11/) | 2023-06-02 11:53:54 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230602-115014-3dc960-pr3540.aar) |
| :heavy_check_mark: | 3dc96062 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/11/) | 2023-06-02 11:54:51 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230602-115014-3dc960-pr3540.zip) |
| | | | | | | |
| :heavy_check_mark: | b1360033 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/12/) | 2023-06-02 12:02:10 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230602-115955-b13600-pr3540.zip) |
| :heavy_check_mark: | b1360033 | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/12/) | 2023-06-02 12:02:47 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230602-115955-b13600-pr3540.zip) |
| :heavy_check_mark: | b1360033 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/12/) | 2023-06-02 12:04:35 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230602-115951-b13600-pr3540.aar) |
| :heavy_check_mark: | b1360033 | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/13/) | 2023-06-02 12:12:23 | ~12 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/13//consoleText) |
| | | | | | | |
| :heavy_check_mark: | aaf8b662 | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/13/) | 2023-06-06 11:08:27 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-110528-aaf8b6-pr3540.zip) |
| :heavy_check_mark: | aaf8b662 | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/13/) | 2023-06-06 11:09:02 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-110524-aaf8b6-pr3540.aar) |
| :heavy_check_mark: | aaf8b662 | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/13/) | 2023-06-06 11:11:51 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-110524-aaf8b6-pr3540.zip) |
| :heavy_check_mark: | aaf8b662 | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/14/) | 2023-06-06 11:20:46 | ~15 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/14//consoleText) |
| | | | | | | |
| :heavy_check_mark: | a4257319 | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/14/) | 2023-06-06 11:29:45 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-112737-a42573-pr3540.zip) |
| :heavy_check_mark: | a4257319 | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/14/) | 2023-06-06 11:31:21 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-112737-a42573-pr3540.aar) |
| :heavy_check_mark: | a4257319 | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/14/) | 2023-06-06 11:33:55 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-112737-a42573-pr3540.zip) |
| :heavy_multiplication_x: | a4257319 | [#15](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/15/) | 2023-06-06 11:35:54 | ~8 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/15//consoleText) |
| :heavy_check_mark: | a4257319 | [#16](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/16/) | 2023-06-06 14:25:51 | ~16 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/16//consoleText) |
| | | | | | | |
| :x: | 80f51ac8 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/15/) | 2023-06-06 22:40:06 | ~21 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/15/consoleText) |
| :x: | 80f51ac8 | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/15/) | 2023-06-06 22:40:27 | ~43 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/15/consoleText) |
| :x: | 80f51ac8 | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/15/) | 2023-06-06 22:40:34 | ~49 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/15/consoleText) |
| :heavy_multiplication_x: | 80f51ac8 | [#17](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/17/) | 2023-06-06 22:40:46 | ~57 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/17//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f2b6b561 | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/16/) | 2023-06-07 14:15:31 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230607-141301-f2b6b5-pr3540.zip) |
| :heavy_check_mark: | f2b6b561 | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/16/) | 2023-06-07 14:16:50 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230607-141258-f2b6b5-pr3540.aar) |
| :heavy_check_mark: | f2b6b561 | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/16/) | 2023-06-07 14:21:34 | ~8 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230607-141259-f2b6b5-pr3540.zip) |
| :heavy_check_mark: | f2b6b561 | [#18](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/18/) | 2023-06-07 14:27:53 | ~14 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/18//consoleText) |
| | | | | | | |
| :heavy_check_mark: | da2bf49d | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/17/) | 2023-06-07 14:30:20 | ~12 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230607-141809-da2bf4-pr3540.zip) |
| :heavy_check_mark: | da2bf49d | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/17/) | 2023-06-07 14:30:50 | ~12 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230607-141805-da2bf4-pr3540.aar) |
| :heavy_multiplication_x: | da2bf49d | [#19](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/19/) | 2023-06-07 14:58:43 | ~30 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/19//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 4f824532 | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/18/) | 2023-06-07 15:35:14 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230607-153257-4f8245-pr3540.zip) |
| :heavy_check_mark: | 4f824532 | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/18/) | 2023-06-07 15:36:26 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230607-153253-4f8245-pr3540.aar) |
| :heavy_multiplication_x: | 4f824532 | [#20](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/20/) | 2023-06-07 15:41:24 | ~8 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/20//consoleText) |
| :heavy_check_mark: | 4f824532 | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/18/) | 2023-06-07 15:41:29 | ~8 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230607-153257-4f8245-pr3540.zip) |
| | | | | | | |
| :heavy_check_mark: | 23c280b9 | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/19/) | 2023-06-08 08:59:55 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230608-085728-23c280-pr3540.zip) |
| :heavy_check_mark: | 23c280b9 | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/19/) | 2023-06-08 09:02:01 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230608-085727-23c280-pr3540.aar) |
| :heavy_check_mark: | 23c280b9 | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/19/) | 2023-06-08 09:07:39 | ~10 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230608-085727-23c280-pr3540.zip) |
| :heavy_multiplication_x: | 23c280b9 | [#21](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/21/) | 2023-06-08 09:20:53 | ~23 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/21//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | f7c8d49b | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/20/) | 2023-06-08 09:05:40 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230608-085959-f7c8d4-pr3540.zip) |
| :heavy_check_mark: | f7c8d49b | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/20/) | 2023-06-08 09:10:21 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230608-090206-f7c8d4-pr3540.aar) |
| :heavy_multiplication_x: | f7c8d49b | [#22](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/22/) | 2023-06-08 09:45:30 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/22//consoleText) |
| :heavy_multiplication_x: | f7c8d49b | [#23](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/23/) | 2023-06-08 10:18:27 | ~29 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/23//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 17b3b78d | [#21](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3540/21/) | 2023-06-08 10:08:10 | ~8 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230608-095936-17b3b7-pr3540.zip) |
| :heavy_check_mark: | 17b3b78d | [#21](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3540/21/) | 2023-06-08 10:12:55 | ~13 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230608-095936-17b3b7-pr3540.zip) |
| :heavy_check_mark: | 17b3b78d | [#21](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3540/21/) | 2023-06-08 10:13:51 | ~14 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230608-095936-17b3b7-pr3540.aar) |
| :heavy_multiplication_x: | 17b3b78d | [#24](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/24/) | 2023-06-08 10:43:03 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/24//consoleText) |
| :heavy_multiplication_x: | 17b3b78d | [#25](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/25/) | 2023-06-08 11:33:30 | ~22 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/25//consoleText) |
| :heavy_check_mark: | 17b3b78d | [#26](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/26/) | 2023-06-08 11:51:29 | ~12 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3540/26//consoleText) |

stefandunca commented on services/wallet/activity/TODO.md:
```diff
@@ -0,0 +1,157 @@
+# DB extension and refactoring for activity
```
This is part of the bigger work package. I will remove it after finishing the filtering work.

Samyoul replied:
This is a nice document @stefandunca

IvanBelyakoff commented on services/wallet/activity/TODO.md:
```diff
@@ -0,0 +1,157 @@
+    - Faster to query (don't know the numbers)
+    - Simpler query will decrease maintenance
+  - Cons:
+    - have to migrate all data, extract and fill the activity on every download of updates for all activities
```
which basically means maintenance complexity and possibly more fragmentation and inconsistency to sum it up. Not arguing which is better, just worried. I feel like we are trading one complexity for another one.

IvanBelyakoff commented on services/wallet/activity/activity.go:
```diff
@@ -442,13 +441,13 @@ func GetActivityEntries(db *sql.DB, addresses []eth.Address, chainIDs []common.C
 			activityStatus := FinalizedAS
 			activityType, filteredAddress := getActivityType(dbTrType)
-			entry = NewActivityEntryWithTransaction(SimpleTransactionPT,
+			entry = NewActivityEntryWithTransaction(false,
```
Honestly I favor the former approach, because you always know what the parameter means when you see the function call. Using 'false' and 'true' at the moment of the call is not readable at all, unless the method is self-explanatory like `setActive(false)`. Otherwise you always guess and check the signature to know what it is.
WDYT?

stefandunca replied:
Fixed; please have another look.

IvanBelyakoff reviewed (approved):
Good job @stefandunca . I'd still appreciate a minor fix that I mentioned, though not insisting.

The bigger thing that bothers me when I see such PRs is that - do we really need to stick to SQLite that hard to make it so painful to play with JSON and do these complex migration changes?
It feels like that a more native and clear approach would be to store some data, that is not sensistive - like blockchain data - that already has JSON structure in some document DB. If what we need to do is mostly store some data and retrieve it - like receipts, logs and such things, we don't need to think how to map the data to SQL rules and use the data natively. That simplifies code, improves readability and might be even not sacrificing performance due to a key-value type of data in many cases. I'd go for it even now, considering that you have many things to do to proceed with this implementation.
Also @alaibe @saledjenic @dlipicar and whoever sees it, WDYT?

stefandunca commented:
> ... do we really need to stick to SQLite that hard to make it so painful to play with JSON and do these complex migration changes?

If we want to keep the compatibility and not force everyone to start from scratch then yes.

> It feels like that a more native and clear approach would be to store some data, that is not sensistive - like blockchain data - that already has JSON structure in some document DB.

AFAIK there is no nosql DB to tick all the requirements we need, searched a few times in the last years for other projects. Some of the requirements that the current approach has:
- Encryption for privacy with a good performance and without having to think at a micro-level about which property needs privacy and which doesn't disclose any
- Maturity (well known and well documented)
- Mobile fit for low power devices

There was a discussion in the team at the last offsite about upgrading to a more feature-rich DB but it was discarded due to none meeting the requirements 

> If what we need to do is mostly store some data and retrieve it - like receipts, logs and such things, we don't need to think how to map the data to SQL rules and use the data natively. 

Current implementation already provides fetching the json documents in an efficient way but it doesn't provide indexed access for filtering. This is what the current implementation improves upon and extracts the data for future use also.

>  ... data that is not sensitive ...

This is a decision hard to make and the benefits of micro-managing the data will quickly be overwhelmed by the relative simplicity of the current approach.

stefandunca commented:
With the latest push, I updated the implementation to provide better error reporting and fixed an issue with SQLite throwing errors when storing `uint64`

dlipicar commented on appdatabase/database.go:
```diff
@@ -198,11 +188,144 @@ func MigrateEnsUsernames(db *sql.DB) error {
+	// Dropped storing postState because it was replaced by the status after EIP 658
+	// Dropped duplicating logs until we have a better way to store them in a
+	// more structured way. They can be found in transfers.receipt still
+	// Dropped tx_hash because it is redundant with transaction_details.hash
```
Note about this: The `hash` column in the `transfers` table is not (always) the transaction hash. Each entry in that table could either be:
- A native token (ETH) transfer. This has no associated entry in the Event Log.
- An ERC20/ERC721 token transfer. Each of these has an associated entry in the Event Log with the `Transfer` signature. 

A particular transaction (with its unique transaction hash) could either contain
- 1 ETH transfer
- 1/many ERC20/ERC721 transfers.

The `hash` column contains:
- For ETH transfers, the transaction hash. This is the one you can use on Etherscan.
- For ERC20/ERC721 transfers, a hash built by us, composed of the transaction hash and the index of the Event Log entry that originated the transfer.

The "real" transaction hash for each entry is the one contained in the receipt.

stefandunca replied:
Thanks for the heads up @dlipicar. So the hash we have is our own generated has (transfer entry ID). Will add tx_has moving forward.

IvanBelyakoff replied:
I would split ETH and ERC20 transfers into different tables. They are different in nature, ERC20's hash is calculated differently for our need, it has logs and maybe something else that I forgot to mention. I don't think there is a strong need to keep them in the same table, while splitting them may simplify tailoring the tables for specifics of these transfer types and easier migration, extension, maintenance. The drawback would be JOINing tables on queries. And in code we mostly handle them differently as well AFAIK, so since we are refactoring, imo we should split them.

stefandunca replied:
> I would split ETH and ERC20 transfers into different tables.

@IvanBelyakoff I agree that a more extensive reorganization would fit the latest requirements in a better way. However, this PR already does more than its intended scope, so let's pick up this idea after the activity work is finished, which can be served with the changes in this PR.

dlipicar commented on appdatabase/database.go:
```diff
@@ -198,11 +188,144 @@ func MigrateEnsUsernames(db *sql.DB) error {
+	// Extract useful information from the receipt blob and store it as sql interpretable
+	//
+	// Dropped storing postState because it was replaced by the status after EIP 658
+	// Dropped duplicating logs until we have a better way to store them in a
```
The `log` entry points to the log that originated the ERC20/ERC721 transfer. Agree about not duplicating, but if we do not store a copy of the log, we should at least store its index (`view.go` uses it to get the actual information we need about the transfer. IMO this kind of metadata should be stored in a DB -multitransaction?- and not generated on the fly)

stefandunca replied:
> The log entry points to the log that originated the ERC20/ERC721 transfer. Agree about not duplicating, but if we do not store a copy of the log, we should at least store its index (view.go uses it to get the actual information we need about the transfer

Indeed log entry is duplicated; however, it makes sense to migrate the index now. WDYT?

> IMO this kind of metadata should be stored in a DB -multitransaction?- and not generated on the fly)

We discussed refactoring the entire data; however, given the time constraints, I decided that this would be too much work that won't benefit the short-term activity feature and leave it for some further reorganization based on the new requirements later on.

Regarding why no multi-transaction for the new fields is because not all transactions in transfers are recorded in multi-transaction and for the current work there is not much difference.

stefandunca replied:
With the latest changes I migrated also the `log_index` by extracting it from the log column.

dlipicar commented on appdatabase/database.go:
```diff
@@ -198,11 +188,144 @@ func MigrateEnsUsernames(db *sql.DB) error {
+	// Extract useful information from the tx blob
+	// Dropped access_list, need a separate table for it
+	// Already there chain_id, type, hash
+	// Dropped v, r, s because I see no way to be useful as BLOBs
```
Is the plan to eventually drop the tx and receipt columns? Or will they always be available in case we need access to the full data?

stefandunca replied:
Not sure yet. Ideally, a proper reorganization would be required with the main transaction information we use in the `transactions` table (now transfers) and everything else linked to it, including the raw data if we require it for future extension. The rest of the aggregated data is required by the activity in the activity (now multi-transaction) table.

As discussed last time we met, the complete re-organization is more consistent and would be better done without migration and before the wide release.

dlipicar commented on appdatabase/database.go:
```diff
@@ -198,11 +188,144 @@ func MigrateEnsUsernames(db *sql.DB) error {
+	// Dropped tx_hash because it is redundant with transaction_details.hash
+	// Dropped the bloom filter because in SQLite is not possible to use it in an
+	// efficient manner
+	// Extract useful information from the tx blob
```
Another fun fact: The `blk_hash` we use for `eth` transfers is wrong. It gets locally calculated using the transaction header, but it doesn't match the real one. That's the reason why the `blocks` table has a bunch of entries with the same `blk_number` but different `blk_hash`. The real block hash is the one that's in the receipt.
This is getting fixed for new entries in my Swap PR since it was causing issues, but I'm not fixing what's already on the DB.

stefandunca replied:
I extracted the original block hash and added it as a new `tx_hash` column.

dlipicar commented on appdatabase/database.go:
```diff
@@ -198,11 +188,144 @@ func MigrateEnsUsernames(db *sql.DB) error {
+		// insert extracted data into the new columns
+		if len(batchEntries) > 0 {
+			var stmt *sql.Stmt
+			stmt, err = sqlTx.Prepare(`UPDATE transfers SET status = ?, receipt_type = ?, cumulative_gas_used = ?, contract_address = ?, gas_used = ?, tx_index = ?,
```
not sure if by `receipt_type` you meant "type obtained from the receipt" or "type of receipt", I think the proper name for this is "Transaction Type" (https://docs.infura.io/infura/networks/ethereum/concepts/transaction-types)

stefandunca replied:
I mean, the type from the type filed from the receipt `Receipt.Type`

stefandunca replied:
> ... I think the proper name for this is "Transaction Type" (https://docs.infura.io/infura/networks/ethereum/concepts/transaction-types)

My understanding is that the Transaction Type is retrieved by the `Transaction.Type()` call which extracts it from `TxData.txType` and which has a different meaning than the type found in the receipt.

dlipicar commented on appdatabase/database.go:
```diff
@@ -198,11 +188,144 @@ func MigrateEnsUsernames(db *sql.DB) error {
+				gasFeeCap := sqlite.BigIntToClampedInt64(tx.GasFeeCap())
+				value := sqlite.BigIntToPadded128BitsStr(tx.Value())
+
+				currentRow = append(currentRow, tx.Protected(), tx.Gas(), gasPrice, gasTipCap, gasFeeCap, value, tx.Nonce(), int64(tx.Size()))
```
There's a couple other fields I think we won't be able to reconstruct without resorting to the JSON blobs
```
	if t.BaseGasFees != "" {
		baseFee := new(big.Int)
		baseFee.SetString(t.BaseGasFees[2:], 16)
		tip := t.Transaction.EffectiveGasTipValue(baseFee)

		view.EffectiveTip = (*hexutil.Big)(tip)
		price := new(big.Int).Add(baseFee, tip)
		view.EffectiveGasPrice = (*hexutil.Big)(price)
	}
```
```
	view.Input = hexutil.Bytes(t.Transaction.Data())
```
```
		if t.Transaction.To() != nil {
			view.To = *t.Transaction.To()
		}
```

stefandunca replied:
I had a quick look, and this seems to be a task for the extensive DB reorganization we talked about recently that won't require data migration.

dlipicar commented on sqlite/fields_test.go:
```diff
@@ -0,0 +1,59 @@
+			expected: strToPtr("000000000052b7d2dcc80cd2e4000000"),
+		},
+		{
+			name:     "case max",
```
please add a test with a negative number

stefandunca replied:
That is undefined behaviour and should not be attempted. Currently, we don't provide negative values but only convert unsigned data, will add a comment to clarify this.

Samyoul commented:
Before I begin my review of this code I wanted to add my support to your points here https://github.com/status-im/status-go/pull/3540#issuecomment-1568352817 .

And particularly to agree that If we could find a nosql db that does meet the requirements highlighted this is something that we'd happily consider.

Samyoul commented on appdatabase/migrations/sql/1685134628_transfers_receipt_json_blob_out.up.sql:
```diff
@@ -0,0 +1,23 @@
+-- ALTER TABLE transfers ADD COLUMN size INT;
+
+-- CREATE INDEX transfers_status ON transfers (status);
+
```
I'm curious why this SQL migration is done in go code? Normally schema changes are done via migrations. I'm not saying we shouldn't, I am just wondering.

stefandunca replied:
I tried to explain it in the comments here: https://github.com/status-im/status-go/blob/d8b24ca160bd52df99d49dc09a903d3cd6914590/sqlite/migrate.go#L26

The migration API doesn't support rolling back to the previous migration to leave the DB in a consistent state (in case of custom golang migration failing). Therefore, I use migration down to roll back to the previous migration (before the custom one). The `down` migration for the current custom migration can be done in SQL. Still, it is more complicated than a simple `tx.Rollback` in the go code (have to remove rows, create a temporary DB to drop the new columns and recopy data). However, for another type of migration, providing an `up` SQL file and a `down` SQL file that discards changes made by the `up` one, might be a better idea and it will work the same.

Hope this clarifies your concerns.

Samyoul commented on appdatabase/migrations/sql/1685134628_transfers_receipt_json_blob_out.down.sql:
```diff
@@ -0,0 +1 @@
+-- This migration is in GO code. If GO migration failure this serves to rollback to the previous one
```
Just a note about our migrations, we don't ever roll a migration back. This is because we can't control the clients and so have no control over what version the users have.

stefandunca replied:
As mentioned in the previous answer, this is required only in the case of custom migrations as an alternative to the API limitations of the migration system to unroll the changes made by the custom code in case of failures. Hence, we can know if a Golang failed, rolled-back, or wasn't run. Otherwise, the DB migration will mark the current custom version as run after running the SQL "marker" migration. Hence after fixing the custom migration, it won't be rerun, and to fix will have to create another migration. Long story short, it is a feature to make fixing Golang migrations easy and optional.

stefandunca replied:
With this approach, we have a relatively simple and robust way of running Golang migrations along when there is no way around it.

Samyoul replied:
Ah ok, @stefandunca I see. Rollback in the event the go code fails, that makes sense to me. Thank you for your patience in explaining this to me.

Samyoul reviewed (approved):
Looks really good. Thank you for this work @stefandunca

stefandunca commented:
@dlipicar I extracted `token_address`, `token_id` and updated `type` columns in the migration process. Please have another look.
@anastasiyaig could you please retest the latest changes once again?

stefandunca commented:
Rebasing on the latest master fails migration tests, looking into it right now. It worked on the morning master though.
